service-base/src/main/java/org/eclipse/hono/service/http/DefaultFailureHandler.java

```diff
@@ -71,6 +72,8 @@ public void handle(final RoutingContext ctx) {
                     } else if (ctx.failure() instanceof HttpException) {
                         final HttpException e = (HttpException) ctx.failure();
                         sendError(ctx.response(), e.getStatusCode(), e.getMessage());
+                    } else if (ctx.statusCode() >= 400 && ctx.statusCode() < 600) {
+                        sendError(ctx.response(), ctx.statusCode(), ctx.failure().getMessage());
```

Comment:
IMHO we should not put the exception's detail message into the response body. Otherwise we might expose details regarding the internal implementation to clients/attackers. For the same reason we should probably also change the code for handling *ServiceInvocationException*s above and use `ServiceInvocationException.getClientFacingMessage()` instead?

Reply:
Sticking to the methods available in the `ServiceInvocationException` class, we could use 
`ServiceInvocationException.getErrorMessageForExternalClient(ctx.failure())`
here.
This however would only override the `ctx.failure().getMessage()` message in case a `ServerErrorException` was thrown.
For the conditional block added above, this would make no difference.

The `getErrorMessageForExternalClient` method originated from the code that assembled the error message for AMQP adapter responses:
https://github.com/eclipse/hono/pull/2499/files#diff-5af44e2d2a3f6823bd0dfa0304a475f621149fe320d4f57923b08ad10626cf14R1290

It is now used in the CoAP and MQTT adapters as well.
I would adapt that method in a separate PR.

I've adapted the last commit here to use `getErrorMessageForExternalClient` in the `DefaultFailureHandler` now.